### PR TITLE
feat: support retrying failed sync attempts

### PR DESCRIPTION
**Pull request by alexmt**
PR introduces fails sync attempt retries. The retries are supported in both manually triggered sync and auto sync.

Manual sync:

```bash
argocd app sync guestbook --retry-limit 5 # 5 retries with default backoff parameters 5s, 10s, 20s, 40s, 80s delays between attempts
```

Default retry strategy ( including automated sync ):

```yaml
# Sync policy
  syncPolicy:
    retry:
      limit: 5 # number of failed sync attempt retries; unlimited number of attempts if less than 0
      backoff:
        duration: 5s # the amount to back off. Default unit is seconds, but could also be a duration (e.g. "2m", "1h")
        factor: 2 # a factor to multiply the base duration after each failed retry
        maxDuration: 3min # the maximum amount of time allowed for the backoff strategy
```
**Note:**

By default auto-sync uses retry strategy with 5 attempts 5 second delay and factor 2 . I've decided to enable retry by default for auto-sync because this seems to be desired behavior and was not implemented from the beginning only due to lack of time.

### CODE REVIEW

**Comment by jannfis** [comment]
I have tested this feature locally and it didn't work out like I would expect. What I have done is:

* Create an application from argocd-example-apps (guestbook) with auto-sync enabled, and the destination set to a non-existent namespace (`guestbook`)
* Wait for the auto-sync to start after application creation - it will correctly detect error state (non-existing namespace destination) and enter the retry loop
* Create namespace `guestbook` on the cluster

However, after the namespace is created and the next sync attempt is retried, the error condition is not recovered and ArgoCD stays in the retry loop until the last attempt, where it will eventually declare the sync as failed. Once I start a manual sync afterwards, the application syncs fine.

**Comment by alexmt** [comment]
Great catch. Thank you @jannfis . 

The retry attempt failed because previous sync operation results were not cleared correctly. Fixed by https://github.com/argoproj/argo-cd/pull/3997/commits/6c12235feaba9b77f07f4a09593f799a2851e73f 

Please take another look.

**Comment by jannfis** [comment]
Just re-tested, and above use-case now works. However, in autosync, the `retryCount` seems not to be reset after successful sync:

```
$ kubectl get applications.argoproj.io guestbook -o json | jq '.status.operationState.phase'
"Succeeded"
$ kubectl get applications.argoproj.io guestbook -o json | jq '.status.operationState.retryCount'
2
```
Is that intended?

**Comment by jannfis** [comment]
After a successful manual sync, the `retryCount` is correctly removed, btw.

**Comment by alexmt** [comment]
Thanks for testing it @jannfis . Honestly, I did not think carefully about when `retryCount` is supposed to be reset. 

I think this is ok that `retryCount` is not removed after successful sync completion. This allows to programmatically retrieve information about number of attempts happened before all resources got synced.

Next sync attempt (both triggered manually or using auto-sync) reset whole `operationState` field so `retryCount` won't affect next sync operation

File: controller/appcontroller_test.go [hunk 1]
```diff
@@ -1009,3 +1009,91 @@ func TestFinalizeProjectDeletion_DoesNotHaveApplications(t *testing.T) {
 		},
 	}, receivedPatch)
 }
+
+func TestProcessRequestedAppOperation_FailedNoRetries(t *testing.T) {
```
Comment (jannfis) [inline controller/appcontroller_test.go:1013]: :+1:

**Review by jannfis — APPROVED** [review]
LGTM

**Pull request merged**